### PR TITLE
[Scenes] Mode Select Cluster Handler

### LINKED ISSUE #11745: Mode Select Cluster does not support scene
#### Problem
 * Mode Select Cluster's CurrentMode attribute should support Scene, as specified by the Specification
 * CurrentMode Attribute is not supporting scene

 #### Proposed Solution
Add implementation in `src/app/clusters/scenes/scenes.cpp` to support CurrentMode

### PR DESCRIPTION
**Pull request by lpbeliveau-silabs**
Added default handler to allow Scenes to interact with the mode select cluster.
Fixes : https://github.com/project-chip/connectedhomeip/issues/11745

### CODE REVIEW

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 1]
```diff
@@ -93,26 +97,183 @@ CHIP_ERROR ModeSelectAttrAccess::Read(const ConcreteReadAttributePath & aPath, A
...
+
+        size_t attributeCount = 0;
+        ReturnErrorOnFailure(attributeValueList.ComputeSize(&attributeCount));
+        VerifyOrReturnError(attributeCount <= scenableAttributeCount, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+        auto pair_iterator = attributeValueList.begin();
+        while (pair_iterator.Next())
+        {
+            auto & decodePair = pair_iterator.GetValue();
+            VerifyOrReturnError(decodePair.attributeID == Attributes::CurrentMode::Id, CHIP_ERROR_INVALID_ARGUMENT);
```
[the remark this answers is not in the document]
Comment (lpbeliveau-silabs) [reply src/app/clusters/mode-select-server/mode-select-server.cpp:212]: In this case, the actions to take in this loop would really depend on what attribute is added (See Color Control and Level Control Handlers).

**Comment by lpbeliveau-silabs** [comment]
@jmartinez-silabs When you have a second, can you validate the fix I made to in attribute-storage, it seemed like there was no way to avoid failure if the number of server == target endpoint

**Comment by lpbeliveau-silabs** [comment]
The fix to attribute-storage will be move to a separate PR once it is validated.

File: src/app/util/attribute-storage.cpp [hunk 2]
```diff
@@ -883,7 +883,7 @@ uint16_t emberAfGetClusterServerEndpointIndex(EndpointId endpoint, ClusterId clu
         }
 
         // If this asserts, the provided fixedClusterServerEndpointCount doesn't match the app data model.
-        VerifyOrDie(adjustedEndpointIndex < fixedClusterServerEndpointCount);
+        VerifyOrDie(adjustedEndpointIndex <= fixedClusterServerEndpointCount);
```
Comment (jmartinez-silabs) [inline src/app/util/attribute-storage.cpp:886]: This doesn't seem right at first glance. The fix endpoints index are from 0 to (fix endpoints count -1)

If you have one fix endpoint you can conly have index 0. If you have 2 fixed endpoint your index are 0 or 1, etc

Comment (lpbeliveau-silabs) [reply src/app/util/attribute-storage.cpp]: Right, this was not the right fix.

**Review by jmartinez-silabs — APPROVED** [review]

**Comment by andy31415** [comment]
@lpbeliveau-silabs @jmartinez-silabs please provide feedback if this is stale or we should still merge it. Doing a review of old PRs for stale closing.

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 3]
```diff
@@ -109,35 +114,201 @@ CHIP_ERROR ModeSelectAttrAccess::Read(const ConcreteReadAttributePath & aPath, A
...
+    static constexpr uint8_t scenableAttributeCount = 1;
+
+    DefaultModeSelectSceneHandler() = default;
+    ~DefaultModeSelectSceneHandler() override {}
+
+    // Default function for the mode select cluster, only puts the mode select cluster ID in the span if supported on the given
+    // endpoint
+    virtual void GetSupportedClusters(EndpointId endpoint, Span<ClusterId> & clusterBuffer) override
+    {
+        ClusterId * buffer = clusterBuffer.data();
```
Comment (andy31415) [inline src/app/clusters/mode-select-server/mode-select-server.cpp:168]: you can move this below. Also you probably can do directly `clusterBuffer[0] = ...` without referencing data.

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 4]
```diff
@@ -109,35 +114,201 @@ CHIP_ERROR ModeSelectAttrAccess::Read(const ConcreteReadAttributePath & aPath, A
...
+        return checkSupportedModeStatus;
     }
     ModeSelect::Attributes::CurrentMode::Set(endpointId, newMode);
 
+    return Status::Success;
+}
+
+} // anonymous namespace
+
+#if defined(MATTER_DM_PLUGIN_SCENES_MANAGEMENT) && CHIP_CONFIG_SCENES_USE_DEFAULT_HANDLERS
```
Comment (andy31415) [inline src/app/clusters/mode-select-server/mode-select-server.cpp:143]: could we find a way to limit ifdef code management? Can this be part of a build system config that compiles or does not compile specific parts of the code?

Comment (lpbeliveau-silabs) [reply src/app/clusters/mode-select-server/mode-select-server.cpp:143]: `CHIP_CONFIG_SCENES_USE_DEFAULT_HANDLERS ` Could be part of a build config, but it is currently being used for all the other cluster's handlers as well so I'd rather move this to another PR.

Essentially the scenes cluster needs this define to work out of the box, so I could build a : `use_custom_scenes_handler` build flag to set this config to false.

Comment (lpbeliveau-silabs) [reply src/app/clusters/mode-select-server/mode-select-server.cpp:143]: Opened issue here: https://github.com/project-chip/connectedhomeip/issues/36102

**Review by andy31415 — APPROVED** [review]

**Pull request merged**